File: src/main/java/me/oriharel/machinery/MaterialChance.java
package me.oriharel.machinery;

import org.bukkit.Material;

import java.util.*;
import java.util.stream.Collectors;

public final class MaterialChance {
    private final List<Material> materials;
    private final int minimumAmount;
    private final int maximumAmount;

    public MaterialChance(List<Material> materials, int minimumAmount, int maximumAmount) {
        this.materials = materials;
        this.minimumAmount = minimumAmount;
        this.maximumAmount = maximumAmount;
    }

    public MaterialChance(int minimumAmount, int maximumAmount, Material... materials) {
        this.materials = Arrays.stream(materials).collect(Collectors.toList());
        this.minimumAmount = minimumAmount;
        this.maximumAmount = maximumAmount;
    }

    public MaterialChance(Material material, int minimumAmount, int maximumAmount) {
        this.materials = new ArrayList<>(Collections.singleton(material));
        this.minimumAmount = minimumAmount;
        this.maximumAmount = maximumAmount;
    }

    public MaterialChance(int minimumAmount, int maximumAmount) {
        this.materials = new ArrayList<>();
        this.minimumAmount = minimumAmount;
        this.maximumAmount = maximumAmount;
    }

    public List<Material> getMaterials() {
        return materials;
    }

    public int getMinimumAmount() {
        return minimumAmount;
    }

    public int getMaximumAmount() {
        return maximumAmount;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        MaterialChance that = (MaterialChance) o;
        return
                minimumAmount == that.minimumAmount &&
                        maximumAmount == that.maximumAmount &&
                        materials == that.materials;
    }

    @Override
    public int hashCode() {
        return Objects.hash(materials, minimumAmount, maximumAmount);
    }

    @Override
    public String toString() {
        return "MaterialChance{" +
                "material=" + materials +
                ", minimumAmount=" + minimumAmount +
                ", maximumAmount=" + maximumAmount +
                '}';
    }
}

File: src/main/kotlin/me/oriharel/machinery/resources/MaterialChance.kt
package me.oriharel.machinery.resources

import com.mojang.datafixers.util.Pair
import me.oriharel.machinery.machines.machine.MachineResourceGetProcess
import me.oriharel.machinery.message.ChanceableOperation
import me.oriharel.machinery.resources.chance.Range
import org.bukkit.Material
import org.bukkit.inventory.ItemStack
import java.util.stream.Collectors

class MaterialChance(private val materials: List<Pair<Material?, Range>>) : ChanceableOperation<List<ItemStack?>?, MachineResourceGetProcess?> {
    override fun getChanced(lootModifier: Double): List<ItemStack?> {
        return materials.stream().map { p: Pair<Material?, Range> -> ItemStack(p.first!!, (p.second.random() * lootModifier).toInt()) }.collect(Collectors.toList())
    }

    override fun executeChanceOperation(param1: MachineResourceGetProcess?, lootModifier: Double) {
        param1?.itemsGained?.addAll(getChanced(lootModifier))
    }

}